File: wrappers/python/asset_identifiers/registry.py
#!/usr/bin/env python3
"""
Asset Identifier Registry — main wrapper class.

Loads identifiers.json and provides lookup methods for
ISIN, CUSIP, SEDOL, FIGI, LEI, and ticker+exchange.

This wrapper is dependency-free and uses only the Python
standard library. It is designed to match the API pattern
of the ISO 4217 and Exchange Calendar registry wrappers.

Example:
    from asset_identifiers import AssetRegistry

    registry = AssetRegistry("identifiers.json")

    # Look up by ISIN
    aapl = registry.by_isin("US0378331005")
    print(aapl["ticker"])  # AAPL

    # Look up by ticker on a specific exchange
    pru_nyse = registry.by_ticker("PRU", "XNYS")
    print(pru_nyse["isin"])  # US7443201022

    # Look up by ticker without exchange (returns list)
    pru_all = registry.by_ticker("PRU")
    print(len(pru_all))  # 2

    # Iterate over all instruments
    for instrument in registry:
        print(instrument["ticker"])
"""

import json
from pathlib import Path
from typing import Dict, Iterator, List, Optional, Union


class AssetRegistry:
    """
    Asset Identifier Registry wrapper.

    Loads a JSON registry of financial instruments and provides
    lookup methods for various identifier types.

    The registry is immutable after loading. If the underlying
    JSON file changes, create a new AssetRegistry instance.

    Attributes:
        count: Number of instruments in the registry
        instruments: List of all instrument dicts
    """

    def __init__(self, path: Union[str, Path] = "identifiers.json"):
        """
        Initialize the registry.

        Args:
            path: Path to identifiers.json. Accepts str or Path.

        Raises:
            FileNotFoundError: If the path does not exist
            json.JSONDecodeError: If the file is not valid JSON
        """
        self._path = Path(path)
        self._load()

    def _load(self) -> None:
        """Load the registry from disk into memory."""
        with open(self._path, "r", encoding="utf-8") as f:
            self._data = json.load(f)

        self._instruments: List[Dict] = self._data.get("instruments", [])
        self._meta: Dict = self._data.get("meta", {})

        # Build lookup indices for O(1) access
        self._isin_index: Dict[str, Dict] = {}
        self._cusip_index: Dict[str, Dict] = {}
        self._sedol_index: Dict[str, Dict] = {}
        self._figi_index: Dict[str, Dict] = {}
        self._lei_index: Dict[str, Dict] = {}
        self._ticker_index: Dict[str, List[Dict]] = {}
        self._exchange_index: Dict[str, List[Dict]] = {}
        self._asset_class_index: Dict[str, List[Dict]] = {}
        self._country_index: Dict[str, List[Dict]] = {}
        self._currency_index: Dict[str, List[Dict]] = {}

        for inst in self._instruments:
            isin = inst.get("isin")
            if isin:
                self._isin_index[isin] = inst

            cusip = inst.get("cusip")
            if cusip:
                self._cusip_index[cusip] = inst

            sedol = inst.get("sedol")
            if sedol:
                self._sedol_index[sedol] = inst

            figi = inst.get("figi")
            if figi:
                self._figi_index[figi] = inst

            lei = inst.get("lei")
            if lei:
                self._lei_index[lei] = inst

            ticker = inst.get("ticker")
            if ticker:
                if ticker not in self._ticker_index:
                    self._ticker_index[ticker] = []
                self._ticker_index[ticker].append(inst)

            exchange = inst.get("exchange")
            if exchange:
                if exchange not in self._exchange_index:
                    self._exchange_index[exchange] = []
                self._exchange_index[exchange].append(inst)

            asset_class = inst.get("asset_class")
            if asset_class:
                if asset_class not in self._asset_class_index:
                    self._asset_class_index[asset_class] = []
                self._asset_class_index[asset_class].append(inst)

            country = inst.get("country")
            if country:
                if country not in self._country_index:
                    self._country_index[country] = []
                self._country_index[country].append(inst)

            currency = inst.get("currency")
            if currency:
                if currency not in self._currency_index:
                    self._currency_index[currency] = []
                self._currency_index[currency].append(inst)

    # ─── Properties ──────────────────────────────────────────────────

    @property
    def count(self) -> int:
        """Number of instruments in the registry."""
        return len(self._instruments)

    @property
    def instruments(self) -> List[Dict]:
        """List of all instrument dicts."""
        return self._instruments

    @property
    def path(self) -> Path:
        """Path to the registry file."""
        return self._path

    # ─── Metadata ────────────────────────────────────────────────────

    def meta(self) -> Dict:
        """Return registry metadata."""
        return self._meta

    def version(self) -> str:
        """Return registry version."""
        return self._meta.get("version", "unknown")

    def generated(self) -> str:
        """Return registry generation date."""
        return self._meta.get("generated", "unknown")

    def sources(self) -> List[str]:
        """Return list of data sources."""
        return self._meta.get("sources", [])

    # ─── Bulk Access ─────────────────────────────────────────────────

    def all(self) -> List[Dict]:
        """Return all instruments as a list."""
        return self._instruments

    def __iter__(self) -> Iterator[Dict]:
        """Iterate over all instruments."""
        return iter(self._instruments)

    def __len__(self) -> int:
        """Number of instruments."""
        return self.count

    def __repr__(self) -> str:
        """String representation."""
        return f"AssetRegistry(count={self.count}, path='{self._path}')"

    def __str__(self) -> str:
        """Human-readable representation."""
        return (
            f"Asset Identifier Registry v{self.version()} "
            f"({self.count} instruments)"
        )

    # ─── Lookup by Identifier ────────────────────────────────────────

    def by_isin(self, isin: str) -> Optional[Dict]:
        """
        Look up an instrument by ISIN.

        Args:
            isin: 12-character ISIN (e.g., US0378331005)

        Returns:
            Instrument dict or None if not found
        """
        return self._isin_index.get(isin.upper())

    def by_cusip(self, cusip: str) -> Optional[Dict]:
        """
        Look up an instrument by CUSIP.

        Args:
            cusip: 9-character CUSIP (e.g., 037833100)

        Returns:
            Instrument dict or None if not found
        """
        return self._cusip_index.get(cusip.upper())

    def by_sedol(self, sedol: str) -> Optional[Dict]:
        """
        Look up an instrument by SEDOL.

        Args:
            sedol: 7-character SEDOL (e.g., 2046251)

        Returns:
            Instrument dict or None if not found
        """
        return self._sedol_index.get(sedol.upper())

    def by_figi(self, figi: str) -> Optional[Dict]:
        """
        Look up an instrument by FIGI.

        Args:
            figi: 12-character FIGI (e.g., BBG000B9XRY4)

        Returns:
            Instrument dict or None if not found
        """
        return self._figi_index.get(figi.upper())

    def by_lei(self, lei: str) -> Optional[Dict]:
        """
        Look up an instrument by LEI.

        Args:
            lei: 20-character LEI

        Returns:
            Instrument dict or None if not found
        """
        return self._lei_index.get(lei.upper())

    # ─── Lookup by Ticker ────────────────────────────────────────────

    def by_ticker(
        self,
        ticker: str,
        exchange: Optional[str] = None,
    ) -> Union[Optional[Dict], List[Dict]]:
        """
        Look up instruments by ticker symbol.

        Args:
            ticker: Ticker symbol (e.g., AAPL, PRU)
            exchange: Optional MIC code to disambiguate (e.g., XNAS)

        Returns:
            If exchange is provided: Instrument dict or None
            If exchange is not provided: List of matching instruments

        Examples:
            # Single result (ticker is unique)
            registry.by_ticker("AAPL")  # Returns list with 1 item

            # Disambiguated by exchange
            registry.by_ticker("PRU", "XNYS")  # Returns dict
            registry.by_ticker("PRU", "XLON")  # Returns dict

            # Multiple results (ticker is ambiguous)
            registry.by_ticker("PRU")  # Returns list with 2 items
        """
        ticker = ticker.upper()
        matches = self._ticker_index.get(ticker, [])

        if exchange is not None:
            exchange = exchange.upper()
            for inst in matches:
                if inst.get("exchange") == exchange:
                    return inst
            return None

        return matches

    # ─── Filtering ───────────────────────────────────────────────────

    def by_exchange(self, exchange: str) -> List[Dict]:
        """
        Return all instruments listed on a given exchange.

        Args:
            exchange: MIC code (e.g., XNAS)

        Returns:
            List of instrument dicts
        """
        return self._exchange_index.get(exchange.upper(), [])

    def by_asset_class(self, asset_class: str) -> List[Dict]:
        """
        Return all instruments of a given asset class.

        Args:
            asset_class: Asset class (equity, etf, bond, etc.)

        Returns:
            List of instrument dicts
        """
        return self._asset_class_index.get(asset_class.lower(), [])

    def by_country(self, country: str) -> List[Dict]:
        """
        Return all instruments from a given country.

        Args:
            country: ISO 3166-1 alpha-2 country code (e.g., US)

        Returns:
            List of instrument dicts
        """
        return self._country_index.get(country.upper(), [])

    def by_currency(self, currency: str) -> List[Dict]:
        """
        Return all instruments trading in a given currency.

        Args:
            currency: ISO 4217 currency code (e.g., USD)

        Returns:
            List of instrument dicts
        """
        return self._currency_index.get(currency.upper(), [])

    # ─── Aggregate Information ───────────────────────────────────────

    def exchanges(self) -> List[str]:
        """
        Return sorted list of all exchanges in the registry.

        Returns:
            List of MIC codes, sorted alphabetically
        """
        return sorted(self._exchange_index.keys())

    def asset_classes(self) -> List[str]:
        """
        Return sorted list of all asset classes in the registry.

        Returns:
            List of asset class strings, sorted alphabetically
        """
        return sorted(self._asset_class_index.keys())

    def currencies(self) -> List[str]:
        """
        Return sorted list of all currencies in the registry.

        Returns:
            List of ISO 4217 currency codes, sorted alphabetically
        """
        return sorted(self._currency_index.keys())

    def countries(self) -> List[str]:
        """
        Return sorted list of all countries in the registry.

        Returns:
            List of ISO 3166-1 alpha-2 country codes, sorted
        """
        return sorted(self._country_index.keys())

    # ─── Convenience Methods ─────────────────────────────────────────

    def ticker_exists(self, ticker: str, exchange: Optional[str] = None) -> bool:
        """
        Check if a ticker exists in the registry.

        Args:
            ticker: Ticker symbol
            exchange: Optional MIC code

        Returns:
            True if the ticker exists, False otherwise
        """
        if exchange is not None:
            return self.by_ticker(ticker, exchange) is not None
        return len(self.by_ticker(ticker)) > 0

    def isin_exists(self, isin: str) -> bool:
        """
        Check if an ISIN exists in the registry.

        Args:
            isin: ISIN code

        Returns:
            True if the ISIN exists, False otherwise
        """
        return self.by_isin(isin) is not None

    def resolve(
        self,
        identifier: str,
        exchange: Optional[str] = None,
    ) -> Union[Optional[Dict], List[Dict]]:
        """
        Resolve any identifier type to an instrument.

        Automatically detects the identifier type:
        - 12 chars → ISIN
        - 9 chars → CUSIP
        - 7 chars → SEDOL
        - 12 chars starting with BBG → FIGI
        - 20 chars → LEI
        - Otherwise → ticker

        Args:
            identifier: Any identifier string
            exchange: Optional MIC code for ticker disambiguation

        Returns:
            Instrument dict, list of dicts, or None
        """
        identifier = identifier.upper().strip()

        if len(identifier) == 12 and identifier.startswith("BBG"):
            return self.by_figi(identifier)
        elif len(identifier) == 12:
            return self.by_isin(identifier)
        elif len(identifier) == 9:
            return self.by_cusip(identifier)
        elif len(identifier) == 7:
            return self.by_sedol(identifier)
        elif len(identifier) == 20:
            return self.by_lei(identifier)
        else:
            return self.by_ticker(identifier, exchange)

    # ─── Statistical Methods ─────────────────────────────────────────

    def identifier_coverage(self) -> Dict[str, Dict[str, int]]:
        """
        Return identifier coverage statistics.

        Returns:
            Dict with coverage counts and percentages for each
            identifier type: ISIN, CUSIP, SEDOL, FIGI, LEI
        """
        total = self.count
        stats = {}

        for id_type in ["isin", "cusip", "sedol", "figi", "lei"]:
            covered = sum(1 for i in self._instruments if i.get(id_type))
            stats[id_type] = {
                "covered": covered,
                "total": total,
                "percentage": round(100 * covered / total, 2) if total > 0 else 0,
            }

        return stats

    def tickers_with_multiple_listings(self) -> List[str]:
        """
        Return tickers that appear on multiple exchanges.

        Returns:
            List of ticker symbols that have multiple exchange listings
        """
        return [
            ticker for ticker, instruments in self._ticker_index.items()
            if len(instruments) > 1
        ]

File: wrappers/go/registry.go
// Package assetidentifiers provides a lightweight interface to the
// canonical asset identifier registry (identifiers.json).
//
// Example:
//
//	registry, err := assetidentifiers.LoadRegistry("identifiers.json")
//	if err != nil {
//		log.Fatal(err)
//	}
//
//	aapl, ok := registry.ByIsin("US0378331005")
//	if ok {
//		fmt.Println(aapl.Ticker) // AAPL
//	}
package assetidentifiers

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// ─── Error Types ─────────────────────────────────────────────────────

// RegistryError represents an error that occurred while loading
// or querying the registry.
type RegistryError struct {
	Op  string // Operation that failed
	Err error  // Underlying error
}

func (e *RegistryError) Error() string {
	return fmt.Sprintf("asset-identifiers: %s: %v", e.Op, e.Err)
}

func (e *RegistryError) Unwrap() error {
	return e.Err
}

// ─── Data Types ──────────────────────────────────────────────────────

// AssetClass represents the type of financial instrument.
type AssetClass string

// Asset class constants.
const (
	AssetClassEquity AssetClass = "equity"
	AssetClassEtf    AssetClass = "etf"
	AssetClassBond   AssetClass = "bond"
	AssetClassOption AssetClass = "option"
	AssetClassFuture AssetClass = "future"
	AssetClassOther  AssetClass = "other"
)

// ListingStatus represents the status of an exchange listing.
type ListingStatus string

// Listing status constants.
const (
	ListingStatusPrimary          ListingStatus = "PRIMARY"
	ListingStatusSecondary        ListingStatus = "SECONDARY"
	ListingStatusDepositaryReceipt ListingStatus = "DEPOSITARY_RECEIPT"
	ListingStatusDelisted         ListingStatus = "DELISTED"
)

// CorporateActionType represents the type of corporate action.
type CorporateActionType string

// Corporate action type constants.
const (
	CorporateActionSplit         CorporateActionType = "SPLIT"
	CorporateActionReverseSplit  CorporateActionType = "REVERSE_SPLIT"
	CorporateActionDividend      CorporateActionType = "DIVIDEND"
	CorporateActionSpinoff       CorporateActionType = "SPINOFF"
	CorporateActionMerger        CorporateActionType = "MERGER"
	CorporateActionAcquisition   CorporateActionType = "ACQUISITION"
	CorporateActionRightsIssue   CorporateActionType = "RIGHTS_ISSUE"
	CorporateActionBuyback       CorporateActionType = "BUYBACK"
)

// ChangeType represents the type of ticker change event.
type ChangeType string

// Ticker change type constants.
const (
	ChangeTypeNone          ChangeType = "none"
	ChangeTypeRename        ChangeType = "rename"
	ChangeTypeDelisting     ChangeType = "delisting"
	ChangeTypeRelisting     ChangeType = "relisting"
	ChangeTypeMerger        ChangeType = "merger"
	ChangeTypeAcquisition   ChangeType = "acquisition"
	ChangeTypeSpinoff       ChangeType = "spinoff"
	ChangeTypeSplit         ChangeType = "split"
	ChangeTypeReverseSplit  ChangeType = "reverse_split"
)

// ExchangeListing represents a listing on a specific exchange.
type ExchangeListing struct {
	Exchange      string        `json:"exchange"`
	Ticker        string        `json:"ticker"`
	Currency      string        `json:"currency"`
	Status        ListingStatus `json:"status"`
	ListingDate   *string       `json:"listing_date,omitempty"`
	DelistingDate *string       `json:"delisting_date,omitempty"`
}

// HistoryEvent represents a ticker change event.
type HistoryEvent struct {
	Ticker     string     `json:"ticker"`
	ChangeDate string     `json:"change_date"`
	ChangeType ChangeType `json:"change_type"`
	Reason     *string    `json:"reason,omitempty"`
	Source     *string    `json:"source,omitempty"`
	SourceURL  *string    `json:"source_url,omitempty"`
}

// CorporateAction represents a corporate action event.
type CorporateAction struct {
	Date       string              `json:"date"`
	ActionType CorporateActionType `json:"action_type"`
	Ratio      *string             `json:"ratio,omitempty"`
	Details    *string             `json:"details,omitempty"`
	Source     *string             `json:"source,omitempty"`
	SourceURL  *string             `json:"source_url,omitempty"`
}

// Instrument represents a financial instrument.
type Instrument struct {
	Isin             string             `json:"isin"`
	Cusip            *string            `json:"cusip,omitempty"`
	Sedol            *string            `json:"sedol,omitempty"`
	Figi             *string            `json:"figi,omitempty"`
	Lei              *string            `json:"lei,omitempty"`
	Ticker           string             `json:"ticker"`
	Exchange         string             `json:"exchange"`
	Name             string             `json:"name"`
	Currency         string             `json:"currency"`
	AssetClass       AssetClass         `json:"asset_class"`
	InstrumentType   *string            `json:"instrument_type,omitempty"`
	Sector           *string            `json:"sector,omitempty"`
	Industry         *string            `json:"industry,omitempty"`
	Country          *string            `json:"country,omitempty"`
	Active           bool               `json:"active"`
	ListingDate      *string            `json:"listing_date,omitempty"`
	DelistingDate    *string            `json:"delisting_date,omitempty"`
	Listings         []ExchangeListing  `json:"listings,omitempty"`
	History          []HistoryEvent     `json:"history,omitempty"`
	CorporateActions []CorporateAction  `json:"corporate_actions,omitempty"`
}

// RegistryCoverage represents registry coverage information.
type RegistryCoverage struct {
	Exchanges    []string `json:"exchanges,omitempty"`
	AssetClasses []string `json:"asset_classes,omitempty"`
	Countries    []string `json:"countries,omitempty"`
}

// RegistryMeta represents registry metadata.
type RegistryMeta struct {
	Version        string           `json:"version"`
	Generated      string           `json:"generated"`
	Count          int              `json:"count"`
	Sources        []string         `json:"sources,omitempty"`
	Coverage       RegistryCoverage `json:"coverage,omitempty"`
	BuildTimestamp *string          `json:"build_timestamp,omitempty"`
}

// registryData is the top-level structure of identifiers.json.
type registryData struct {
	Meta        RegistryMeta `json:"meta"`
	Instruments []Instrument `json:"instruments"`
}

// IdentifierCoverageStat represents coverage for a single identifier type.
type IdentifierCoverageStat struct {
	Covered    int     `json:"covered"`
	Total      int     `json:"total"`
	Percentage float64 `json:"percentage"`
}

// IdentifierCoverage represents coverage for all identifier types.
type IdentifierCoverage struct {
	Isin  IdentifierCoverageStat `json:"isin"`
	Cusip IdentifierCoverageStat `json:"cusip"`
	Sedol IdentifierCoverageStat `json:"sedol"`
	Figi  IdentifierCoverageStat `json:"figi"`
	Lei   IdentifierCoverageStat `json:"lei"`
}

// ─── Registry ────────────────────────────────────────────────────────

// AssetRegistry represents the loaded asset identifier registry.
type AssetRegistry struct {
	instruments []Instrument
	meta        RegistryMeta
	path        string

	// Lookup indices
	isinIndex    map[string]int
	cusipIndex   map[string]int
	sedolIndex   map[string]int
	figiIndex    map[string]int
	leiIndex     map[string]int
	tickerIndex  map[string][]int
	exchangeIndex map[string][]int
	assetClassIndex map[string][]int
	countryIndex map[string][]int
	currencyIndex map[string][]int
}

// LoadRegistry loads the registry from a JSON file.
//
// Returns an error if the file cannot be read or parsed, or if
// duplicate identifiers are found.
func LoadRegistry(path string) (*AssetRegistry, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, &RegistryError{Op: "read file", Err: err}
	}

	var data registryData
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, &RegistryError{Op: "parse JSON", Err: err}
	}

	return newRegistry(data, path)
}

// newRegistry creates a registry from parsed data and builds indices.
func newRegistry(data registryData, path string) (*AssetRegistry, error) {
	absPath, _ := filepath.Abs(path)

	registry := &AssetRegistry{
		instruments:     data.Instruments,
		meta:            data.Meta,
		path:            absPath,
		isinIndex:       make(map[string]int, len(data.Instruments)),
		cusipIndex:      make(map[string]int),
		sedolIndex:      make(map[string]int),
		figiIndex:       make(map[string]int),
		leiIndex:        make(map[string]int),
		tickerIndex:     make(map[string][]int),
		exchangeIndex:   make(map[string][]int),
		assetClassIndex: make(map[string][]int),
		countryIndex:    make(map[string][]int),
		currencyIndex:   make(map[string][]int),
	}

	for idx := range data.Instruments {
		inst := &data.Instruments[idx]

		// ISIN
		isin := strings.ToUpper(inst.Isin)
		if _, exists := registry.isinIndex[isin]; exists {
			return nil, &RegistryError{Op: "duplicate ISIN", Err: fmt.Errorf("%s", isin)}
		}
		registry.isinIndex[isin] = idx

		// CUSIP
		if inst.Cusip != nil {
			cusip := strings.ToUpper(*inst.Cusip)
			if _, exists := registry.cusipIndex[cusip]; exists {
				return nil, &RegistryError{Op: "duplicate CUSIP", Err: fmt.Errorf("%s", cusip)}
			}
			registry.cusipIndex[cusip] = idx
		}

		// SEDOL
		if inst.Sedol != nil {
			sedol := strings.ToUpper(*inst.Sedol)
			if _, exists := registry.sedolIndex[sedol]; exists {
				return nil, &RegistryError{Op: "duplicate SEDOL", Err: fmt.Errorf("%s", sedol)}
			}
			registry.sedolIndex[sedol] = idx
		}

		// FIGI
		if inst.Figi != nil {
			figi := strings.ToUpper(*inst.Figi)
			if _, exists := registry.figiIndex[figi]; exists {
				return nil, &RegistryError{Op: "duplicate FIGI", Err: fmt.Errorf("%s", figi)}
			}
			registry.figiIndex[figi] = idx
		}

		// LEI
		if inst.Lei != nil {
			lei := strings.ToUpper(*inst.Lei)
			if _, exists := registry.leiIndex[lei]; exists {
				return nil, &RegistryError{Op: "duplicate LEI", Err: fmt.Errorf("%s", lei)}
			}
			registry.leiIndex[lei] = idx
		}

		// Ticker
		ticker := strings.ToUpper(inst.Ticker)
		registry.tickerIndex[ticker] = append(registry.tickerIndex[ticker], idx)

		// Exchange
		exchange := strings.ToUpper(inst.Exchange)
		registry.exchangeIndex[exchange] = append(registry.exchangeIndex[exchange], idx)

		// Asset class
		assetClass := string(inst.AssetClass)
		registry.assetClassIndex[assetClass] = append(registry.assetClassIndex[assetClass], idx)

		// Country
		if inst.Country != nil {
			country := strings.ToUpper(*inst.Country)
			registry.countryIndex[country] = append(registry.countryIndex[country], idx)
		}

		// Currency
		currency := strings.ToUpper(inst.Currency)
		registry.currencyIndex[currency] = append(registry.currencyIndex[currency], idx)
	}

	return registry, nil
}

// ─── Properties ──────────────────────────────────────────────────────

// Count returns the number of instruments in the registry.
func (r *AssetRegistry) Count() int {
	return len(r.instruments)
}

// Path returns the absolute path to the registry file.
func (r *AssetRegistry) Path() string {
	return r.path
}

// Meta returns the registry metadata.
func (r *AssetRegistry) Meta() *RegistryMeta {
	return &r.meta
}

// Version returns the registry version.
func (r *AssetRegistry) Version() string {
	return r.meta.Version
}

// Generated returns the registry generation date.
func (r *AssetRegistry) Generated() string {
	return r.meta.Generated
}

// Sources returns the list of data sources.
func (r *AssetRegistry) Sources() []string {
	return r.meta.Sources
}

// ─── Bulk Access ─────────────────────────────────────────────────────

// All returns all instruments.
func (r *AssetRegistry) All() []Instrument {
	return r.instruments
}

// Instruments returns a pointer to the instruments slice.
func (r *AssetRegistry) Instruments() *[]Instrument {
	return &r.instruments
}

// ─── Lookup by Identifier ────────────────────────────────────────────

// ByIsin looks up an instrument by ISIN.
// Returns the instrument and true if found.
func (r *AssetRegistry) ByIsin(isin string) (*Instrument, bool) {
	idx, ok := r.isinIndex[strings.ToUpper(isin)]
	if !ok {
		return nil, false
	}
	return &r.instruments[idx], true
}

// ByCusip looks up an instrument by CUSIP.
func (r *AssetRegistry) ByCusip(cusip string) (*Instrument, bool) {
	idx, ok := r.cusipIndex[strings.ToUpper(cusip)]
	if !ok {
		return nil, false
	}
	return &r.instruments[idx], true
}

// BySedol looks up an instrument by SEDOL.
func (r *AssetRegistry) BySedol(sedol string) (*Instrument, bool) {
	idx, ok := r.sedolIndex[strings.ToUpper(sedol)]
	if !ok {
		return nil, false
	}
	return &r.instruments[idx], true
}

// ByFigi looks up an instrument by FIGI.
func (r *AssetRegistry) ByFigi(figi string) (*Instrument, bool) {
	idx, ok := r.figiIndex[strings.ToUpper(figi)]
	if !ok {
		return nil, false
	}
	return &r.instruments[idx], true
}

// ByLei looks up an instrument by LEI.
func (r *AssetRegistry) ByLei(lei string) (*Instrument, bool) {
	idx, ok := r.leiIndex[strings.ToUpper(lei)]
	if !ok {
		return nil, false
	}
	return &r.instruments[idx], true
}

// ─── Lookup by Ticker ────────────────────────────────────────────────

// ByTicker looks up instruments by ticker symbol.
//
// If exchange is non-empty, returns at most one instrument.
// If exchange is empty, returns all matching instruments.
func (r *AssetRegistry) ByTicker(ticker string, exchange string) []*Instrument {
	indices, ok := r.tickerIndex[strings.ToUpper(ticker)]
	if !ok {
		return nil
	}

	if exchange == "" {
		result := make([]*Instrument, 0, len(indices))
		for _, idx := range indices {
			result = append(result, &r.instruments[idx])
		}
		return result
	}

	exchangeUpper := strings.ToUpper(exchange)
	for _, idx := range indices {
		if strings.EqualFold(r.instruments[idx].Exchange, exchangeUpper) {
			return []*Instrument{&r.instruments[idx]}
		}
	}
	return nil
}

// ByTickerExchange looks up a single instrument by ticker and exchange.
func (r *AssetRegistry) ByTickerExchange(ticker string, exchange string) (*Instrument, bool) {
	results := r.ByTicker(ticker, exchange)
	if len(results) == 1 {
		return results[0], true
	}
	return nil, false
}

// ─── Filtering ───────────────────────────────────────────────────────

// ByExchange returns all instruments on a given exchange.
func (r *AssetRegistry) ByExchange(exchange string) []*Instrument {
	indices, ok := r.exchangeIndex[strings.ToUpper(exchange)]
	if !ok {
		return nil
	}
	result := make([]*Instrument, 0, len(indices))
	for _, idx := range indices {
		result = append(result, &r.instruments[idx])
	}
	return result
}

// ByAssetClass returns all instruments of a given asset class.
func (r *AssetRegistry) ByAssetClass(assetClass AssetClass) []*Instrument {
	indices, ok := r.assetClassIndex[string(assetClass)]
	if !ok {
		return nil
	}
	result := make([]*Instrument, 0, len(indices))
	for _, idx := range indices {
		result = append(result, &r.instruments[idx])
	}
	return result
}

// ByCountry returns all instruments from a given country.
func (r *AssetRegistry) ByCountry(country string) []*Instrument {
	indices, ok := r.countryIndex[strings.ToUpper(country)]
	if !ok {
		return nil
	}
	result := make([]*Instrument, 0, len(indices))
	for _, idx := range indices {
		result = append(result, &r.instruments[idx])
	}
	return result
}

// ByCurrency returns all instruments trading in a given currency.
func (r *AssetRegistry) ByCurrency(currency string) []*Instrument {
	indices, ok := r.currencyIndex[strings.ToUpper(currency)]
	if !ok {
		return nil
	}
	result := make([]*Instrument, 0, len(indices))
	for _, idx := range indices {
		result = append(result, &r.instruments[idx])
	}
	return result
}

// ─── Aggregate Information ───────────────────────────────────────────

// Exchanges returns a sorted list of all exchanges.
func (r *AssetRegistry) Exchanges() []string {
	keys := make([]string, 0, len(r.exchangeIndex))
	for k := range r.exchangeIndex {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// AssetClasses returns a sorted list of all asset classes.
func (r *AssetRegistry) AssetClasses() []AssetClass {
	keys := make([]string, 0, len(r.assetClassIndex))
	for k := range r.assetClassIndex {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	classes := make([]AssetClass, 0, len(keys))
	for _, k := range keys {
		classes = append(classes, AssetClass(k))
	}
	return classes
}

// Currencies returns a sorted list of all currencies.
func (r *AssetRegistry) Currencies() []string {
	keys := make([]string, 0, len(r.currencyIndex))
	for k := range r.currencyIndex {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// Countries returns a sorted list of all countries.
func (r *AssetRegistry) Countries() []string {
	keys := make([]string, 0, len(r.countryIndex))
	for k := range r.countryIndex {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// ─── Convenience Methods ─────────────────────────────────────────────

// IsinExists checks if an ISIN exists in the registry.
func (r *AssetRegistry) IsinExists(isin string) bool {
	_, ok := r.ByIsin(isin)
	return ok
}

// TickerExists checks if a ticker exists in the registry.
func (r *AssetRegistry) TickerExists(ticker string, exchange string) bool {
	return len(r.ByTicker(ticker, exchange)) > 0
}

// IdentifierCoverage returns coverage statistics for each identifier type.
func (r *AssetRegistry) IdentifierCoverage() IdentifierCoverage {
	total := r.Count()

	calc := func(covered int) IdentifierCoverageStat {
		percentage := 0.0
		if total > 0 {
			percentage = float64(covered) / float64(total) * 100.0
		}
		return IdentifierCoverageStat{
			Covered:    covered,
			Total:      total,
			Percentage: percentage,
		}
	}

	isinCovered := 0
	cusipCovered := 0
	sedolCovered := 0
	figiCovered := 0
	leiCovered := 0

	for i := range r.instruments {
		if r.instruments[i].Isin != "" {
			isinCovered++
		}
		if r.instruments[i].Cusip != nil {
			cusipCovered++
		}
		if r.instruments[i].Sedol != nil {
			sedolCovered++
		}
		if r.instruments[i].Figi != nil {
			figiCovered++
		}
		if r.instruments[i].Lei != nil {
			leiCovered++
		}
	}

	return IdentifierCoverage{
		Isin:  calc(isinCovered),
		Cusip: calc(cusipCovered),
		Sedol: calc(sedolCovered),
		Figi:  calc(figiCovered),
		Lei:   calc(leiCovered),
	}
}

// TickersWithMultipleListings returns tickers that appear on multiple exchanges.
func (r *AssetRegistry) TickersWithMultipleListings() []string {
	var ambiguous []string
	for ticker, indices := range r.tickerIndex {
		if len(indices) > 1 {
			ambiguous = append(ambiguous, ticker)
		}
	}
	sort.Strings(ambiguous)
	return ambiguous
}

// ─── String Representation ───────────────────────────────────────────

// String returns a human-readable representation of the registry.
func (r *AssetRegistry) String() string {
	return fmt.Sprintf("Asset Identifier Registry v%s (%d instruments)", r.Version(), r.Count())
}